chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/state/AddedToRoom.kt

```diff
@@ -0,0 +1,33 @@
+package com.pusher.chatkit.rooms.state
+
+import com.pusher.chatkit.state.ChatkitState
+import org.reduxkotlin.reducerForActionType
+
+internal data class AddedToRoom(
+    val room: JoinedRoomInternalType,
+    val unreadCount: Int
+)
+
+internal val addedToRoomReducer =
+    reducerForActionType<ChatkitState, AddedToRoom> { state, action ->
+
+            val joinedRoom = Pair(action.room.id, action.room)
+            val joinedRoomUnreadCount = Pair(action.room.id, action.unreadCount)
+            var joinedRoomsState = JoinedRoomsState(
+                    hashMapOf(joinedRoom),
+                    hashMapOf(joinedRoomUnreadCount)
+            )
+
+            if (state.joinedRoomsState != null) {
```

Comment:
the code can be simplified: the state is always there for any subsequent events of subscription

Reply:
Hmm. I see what you're saying but our initial() function means the joinedRoomState can be null - is there anything further we can do to secure ourselves rather than using `!!`?

Reply:
it may be actually more clear and that's the convention to check it explicitly as precondition:
```kotlin
internal val joinedRoomReducer =
    reducerForActionType<ChatkitState, JoinedRoom> { state, action ->
        checkNotNull(state.joinedRoomsState)

        val joinedRoom = action.room.id to action.room
        val joinedRoomUnreadCount = action.room.id to action.unreadCount

        state.with(
            JoinedRoomsState(
                state.joinedRoomsState.rooms + joinedRoom,
                state.joinedRoomsState.unreadCounts + joinedRoomUnreadCount
            )
        )
    }
```
However contracts (that allow smart cast) are a new and still experimental feature in Kotlin 1.3 and only available now for top-level functions. I think it's fine to use them, just worth to know that they may be not applicable everywhere yet.

I have also make a few cosmetic improvements around idiomatic use of the language, conciseness and formatting.

Reply:
👍 that makes sense - have updated all the other reducers to also `checkNotnull` at the start